src/main/java/org/prebid/server/bidder/grid/GridBidder.java

```diff
@@ -1,56 +1,243 @@
+            return imp.toBuilder()
+                    .ext(mapper.mapper().valueToTree(modifiedExtImpGrid))
+                    .build();
+        }
+        return imp;
+    }
 
-            return imp.toBuilder().ext(mapper.mapper().valueToTree(modifiedGridExtImp)).build();
+    private Keywords getKeywordsFromImpExt(JsonNode extImp) {
+        try {
+            return mapper.mapper().convertValue(extImp, GRID_EXT_TYPE_REFERENCE).getBidder().getKeywords();
+        } catch (IllegalArgumentException e) {
+            return null;
         }
+    }
 
-        return imp;
+    private BidRequest modifyRequest(BidRequest bidRequest, Keywords firstImpKeywords, List<Imp> imp) {
+        final User user = bidRequest.getUser();
+        final String userKeywords = user != null ? user.getKeywords() : null;
+        final Site site = bidRequest.getSite();
+        final String siteKeywords = site != null ? site.getKeywords() : null;
+
+        final ExtRequest extRequest = bidRequest.getExt();
+        final Keywords resolvedKeywords = buildBidRequestExtKeywords(
+                ObjectUtils.defaultIfNull(userKeywords, ""),
+                ObjectUtils.defaultIfNull(siteKeywords, ""),
+                firstImpKeywords,
+                getKeywordsFromRequestExt(extRequest));
+
+        return bidRequest.toBuilder()
+                .imp(imp)
+                .ext(modifyExtRequest(extRequest, resolvedKeywords))
+                .build();
+    }
+
+    private Keywords getKeywordsFromRequestExt(ExtRequest extRequest) {
+        try {
+            final JsonNode requestKeywordsNode = extRequest != null ? extRequest.getProperty("keywords") : null;
+            return requestKeywordsNode != null
+                    ? mapper.mapper().treeToValue(requestKeywordsNode, Keywords.class)
+                    : null;
+        } catch (JsonProcessingException e) {
+            return null;
+        }
+    }
+
+    private Keywords buildBidRequestExtKeywords(String userKeywords,
+                                                String siteKeywords,
+                                                Keywords firstImpExtKeywords,
+                                                Keywords requestExtKeywords) {
+        return GridKeywordsUtil.merge(
+                mapper,
+                GridKeywordsUtil.resolveKeywordsFromOpenRtb(userKeywords, siteKeywords, mapper),
+                GridKeywordsUtil.resolveKeywords(firstImpExtKeywords, mapper),
+                GridKeywordsUtil.resolveKeywords(requestExtKeywords, mapper));
+    }
+
+    private ExtRequest modifyExtRequest(ExtRequest extRequest, Keywords keywords) {
+        final boolean extRequestPresent = extRequest != null;
+        final ExtRequestPrebid extRequestPrebid = extRequestPresent ? extRequest.getPrebid() : null;
+        final Map<String, JsonNode> extRequestProperties = extRequestPresent
+                ? extRequest.getProperties()
+                : Collections.emptyMap();
```

Comment:
Is it time to add `org.prebid.server.util.ObjectUtil#getIfNotNullOrDefault` ? ;-)